gravitee-plugin-core/src/main/java/io/gravitee/plugin/core/internal/PluginRegistryImpl.java

```diff
@@ -102,99 +96,101 @@ protected void doStart() throws Exception {
         super.doStart();
 
         if (!init) {
-            LOGGER.info("Initializing plugin registry.");
+            log.info("Initializing plugin registry.");
             this.init();
-            LOGGER.info("Plugins have been loaded and installed.");
+            log.info("Plugins have been loaded and installed.");
         } else {
-            LOGGER.warn("Plugin registry has already been initialized.");
+            log.warn("Plugin registry has already been initialized.");
         }
     }
 
     public void init() throws Exception {
         String[] pluginsPath = configuration.getPluginsPath();
         if ((pluginsPath == null || pluginsPath.length == 0) && workspacesPath == null) {
-            LOGGER.error("No plugin registry configured.");
-            throw new RuntimeException("No plugin registry configured.");
+            log.error("No plugin registry configured.");
+            throw new IllegalArgumentException("No plugin registry configured.");
         }
 
         // Use override configuration
         if (workspacesPath != null) {
             pluginsPath = workspacesPath;
         }
 
-        for (String aWorkspacePath : pluginsPath) {
-            loadPluginsFromRegistry(aWorkspacePath);
-        }
-
-        printPlugins();
-        eventManager.publishEvent(PluginEvent.ENDED, null);
+        Flowable
+            .fromArray(pluginsPath)
+            .flatMap(this::loadPluginsFromPath)
+            // reserve sort
+            .sorted((p1, p2) -> Math.negateExact(((Long) p1.getArchiveTimestamp()).compareTo(p2.getArchiveTimestamp())))
+            // As plugin arrive sorted by reverse file date
+            // we can exclude duplicate
+            .distinct()
+            .doOnNext(plugin -> {
+                plugins.add(plugin);
+                eventManager.publishEvent(PluginEvent.DEPLOYED, plugin);
+            })
+            .doOnComplete(() -> {
+                printPlugins();
```

Comment:
Could you print the plugin on the doOnNext in order to avoid growing the list over again ?

Reply:
refactored as discussed